tools/castle-editor/code/framedesign.pas

```diff
@@ -1776,6 +1786,114 @@ procedure TDesignFrame.CastleControlUpdate(Sender: TObject);
   end;
 end;
 
+procedure TDesignFrame.CastleControlDragOver(Sender, Source: TObject; X,
+  Y: Integer; State: TDragState; var Accept: Boolean);
+var
+  ShellList: TCastleShellListView;
+  SelectedFileName: String;
+  SelectedURL: String;
+begin
+  if Source is TCastleShellListView then
+  begin
+    ShellList := TCastleShellListView(Source);
+    SelectedFileName := ShellList.GetPathFromItem(ShellList.Selected);
+    SelectedURL := FilenameToURISafe(SelectedFileName);
+
+    Accept := TFileFilterList.Matches(LoadScene_FileFilters, SelectedURL);
+  end else
+    Accept := false;
+end;
+
+procedure TDesignFrame.CastleControlDragDrop(Sender, Source: TObject; X, Y: Integer);
+var
+  ShellList: TCastleShellListView;
+  SelectedFileName: String;
+  SelectedURL: String;
+  Scene: TCastleScene;
+  UI: TCastleUserInterface;
+  Viewport: TCastleViewport;
+  ScenePos: TVector3;
+  RayOrigin, RayDirection: TVector3;
+  RayHit: TRayCollision;
+  Distance: Single;
+  OldPickable: Boolean;
+
+  function Is2DAsset(URL : String): Boolean;
+  var
+    MimeType: String;
+  begin
+    MimeType := URIMimeType(URL);
+    Result := LoadImage_FileFilters.Matches(URL) or
+      (MimeType = 'application/json') or
+      (MimeType = 'application/x-starling-sprite-sheet') or
+      (MimeType = 'application/x-plist') or
+      (MimeType = 'application/x-cocos2d-sprite-sheet');
+  end;
+
+begin
+  if Source is TCastleShellListView then
+  begin
+    ShellList := TCastleShellListView(Source);
+    SelectedFileName := ShellList.GetPathFromItem(ShellList.Selected);
+    SelectedURL := MaybeUseDataProtocol(FilenameToURISafe(SelectedFileName));
+
+    if not TFileFilterList.Matches(LoadScene_FileFilters, SelectedURL) then
+      Exit;
+
+    UI := FDesignerLayer.HoverUserInterface(Vector2(X, Y));
+    if not (UI is TCastleViewport) then
+      Exit;
+
+    Viewport := TCastleViewport(UI);
+    try
+      Scene := AddComponent(Viewport.Items, TCastleScene, nil) as TCastleScene;
+
+      if Is2DAsset(SelectedURL) then
+        Scene.Setup2D;
+
+      Scene.URL := SelectedURL;
+
+      { Make gizmos not pickable when looking for new scene position,
+        because ray can hit on gizmo. }
+      OldPickable := VisualizeTransformSelected.Pickable;
+      try
+        VisualizeTransformSelected.Pickable := false;
+        Viewport.PositionToRay(Vector2(X, CastleControl.Height - Y), true, RayOrigin, RayDirection);
+        RayHit := Viewport.Items.WorldRay(RayOrigin, RayDirection);
+      finally
+        VisualizeTransformSelected.Pickable := OldPickable;
+      end;
+      if RayHit <> nil then
+      begin
+        Distance := RayHit.Distance;
+        FreeAndNil(RayHit);
+      end else
+      begin
+        // if we don't hit any other scene set Distance to default
+        if Viewport.Camera.ProjectionType = ptOrthographic then
+        begin
+          // use camera position for 2D (minimum 100 distance)
```

Comment:
I would suggest (but please test in practice) to make a different logic in this case (`Viewport.Camera.ProjectionType = ptOrthographic`, i.e. nothing was hit and we have ortho projection):

Place object at `Z = (Viewport.Camera.EffectiveProjectionNear + Viewport.Camera.EffectiveProjectionFar) / 2` (and this should make Z = 0 in case of default projection set by `TCastleViewport.Setup2D`, which is nice and most natural for user). So you would need to rearrange code, in this case we don't follow distance, but we do 

```pascal
PlaneZ = (Viewport.Camera.EffectiveProjectionNear + Viewport.Camera.EffectiveProjectionFar) / 2;
if not TrySimplePlaneRayIntersection(ScenePos, 2, PlaneZ, RayOrigin, RayDirection) then 
  Exit; // camera direction parallel to 3D plane with Z = constant
```

... and you get ready `ScenePos`, and don't care at all about calculating `Distance`.

Only in other cases, you do `RayOrigin + (RayDirection * Distance)`.